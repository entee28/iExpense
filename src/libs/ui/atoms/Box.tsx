import React from 'react'
import { View, ViewProps, ViewStyle } from 'react-native'

type Props = {
  children?: React.ReactNode
  viewProps?: ViewProps
  testID?: string
} & ViewStyle

export const Box = ({ children, viewProps, testID, ...styles }: Props) => {
  return (
    <View testID={testID} {...viewProps} style={styles}>
      {children}
    </View>
  )
}
