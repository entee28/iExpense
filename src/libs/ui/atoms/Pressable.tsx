import React from 'react'
import {
  ViewStyle,
  TouchableOpacity,
  TouchableOpacityProps
} from 'react-native'

export type PressableProps = Pick<
  TouchableOpacityProps,
  | 'children'
  | 'disabled'
  | 'hitSlop'
  | 'onPress'
  | 'testID'
  | 'activeOpacity'
  | 'style'
  | 'onLongPress'
> &
  ViewStyle

export const Pressable = ({
  activeOpacity,
  children,
  disabled,
  hitSlop,
  onPress,
  testID,
  style,
  onLongPress,
  ...styles
}: PressableProps) => {
  return (
    <TouchableOpacity
      activeOpacity={activeOpacity}
      children={children}
      testID={testID}
      disabled={disabled}
      hitSlop={hitSlop}
      onPress={onPress}
      onLongPress={onLongPress}
      style={[styles, style]}
    />
  )
}
