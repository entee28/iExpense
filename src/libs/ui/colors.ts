export default {
  primary5: '#F7FAFF',
  primary10: '#E5EEFF',
  primary20: '#0055FF33',
  primary30: '#99BBFF',
  primary40: 'rgba(0, 85, 255, 0.4)',
  primary50: 'rgba(0, 85, 255, 0.5)',
  primary80: '#3377FF',
  primary100: '#0055FF',
  mono5: '#1A1A1A0D',
  mono4: '#1A1A1A0A',
  mono10: '#1A1A1A1A',
  mono40: '#1A1A1A66',
  mono60: '#1A1A1A99',
  mono70: '#1A1A1AB2',
  mono80: '#1A1A1ACC',
  mono100: '#1A1A1A',
  warning5: '#FFA8000D',
  warning10: '#FFA8001A',
  warning30: '#FFA8004D',
  warning80: '#FFA80080',
  warning100: '#FFA800',
  warning: '#FF6100',
  destructive: '#E60A32',
  destructive80: '#E60A32CC',
  destructive30: '#E60A324D',
  destructive30Hard: '#F8B6C1',
  destructive10: '#E60A321A',
  destructive5: 'rgba(230, 10, 50, 0.05)',
  positive5: '#00B47E0D',
  positive10: '#00B47E1A',
  positive30: '#00B47E4D',
  positive100: '#00B47E',
  white: '#FFFFFF',
  transparent: 'transparent',
  aliceBlue: '#EBF3FF'
}
