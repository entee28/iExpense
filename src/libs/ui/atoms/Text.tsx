import React, { useMemo } from 'react'
import { TextProps, TextStyle, Text as RNText } from 'react-native'
import colors from '../colors'

type Props = {
  children?: React.ReactNode
  textProps?: TextProps
  extraBold?: boolean
  bold?: boolean
  semiBold?: boolean
  medium?: boolean
  regular?: boolean
} & TextStyle

export enum FontFamily {
  extraBold = 'Nunito-ExtraBold',
  bold = 'Nunito-Bold',
  medium = 'Nunito-Medium',
  semiBold = 'Nunito-SemiBold',
  regular = 'Nunito-Regular'
}

export const Text = ({
  textProps,
  children,
  fontSize = 14,
  lineHeight = 22,
  color = colors.mono100,
  textAlign = 'auto',
  extraBold,
  bold,
  semiBold,
  medium,
  regular,
  ...styles
}: Props) => {
  const fontFamily = useMemo(() => {
    if (extraBold) {
      return FontFamily.extraBold
    }
    if (bold) {
      return FontFamily.bold
    }
    if (semiBold) {
      return FontFamily.semiBold
    }
    if (medium) {
      return FontFamily.medium
    }
    if (regular) {
      return FontFamily.regular
    }
    return FontFamily.regular
  }, [extraBold, bold, semiBold, medium, regular])

  return (
    <RNText
      {...textProps}
      style={[{ color, textAlign, lineHeight, fontSize, fontFamily }, styles]}>
      {children}
    </RNText>
  )
}
