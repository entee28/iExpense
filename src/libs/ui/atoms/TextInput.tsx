import React, { useMemo, useState } from 'react'
import {
  StyleProp,
  TextInputProps,
  TextStyle,
  View,
  ViewStyle,
  TextInput as RNTextInput,
  NativeSyntheticEvent,
  TextInputFocusEventData
} from 'react-native'
import { Box, Text, FontFamily } from '../atoms'
import { InvalidIndicator, ValidIndicator } from 'assets/icons'
import colors from '../colors'

type Props = {
  error?: string
  isValid?: boolean
  containerStyle?: ViewStyle
  inputStyle?: StyleProp<TextStyle>
  label?: string | JSX.Element
  labelTestID?: string
  msgTestID?: string
} & FontWeight &
  TextInputProps &
  Pick<TextStyle, 'fontSize' | 'lineHeight' | 'color'>

const INPUT_PADDING_HORIZONTAL = 16
const INPUT_PADDING_VERTICAL = 12

export const TextInput = ({
  error,
  isValid,
  containerStyle,
  inputStyle,
  editable,
  label,
  bold,
  semiBold,
  medium,
  extraBold,
  regular,
  fontSize,
  lineHeight,
  color,
  onFocus,
  onBlur,
  msgTestID,
  labelTestID,
  ...inputProps
}: Props) => {
  const fontFamily = useMemo(() => {
    if (extraBold) {
      return FontFamily.extraBold
    }
    if (bold) {
      return FontFamily.bold
    }
    if (semiBold) {
      return FontFamily.semiBold
    }
    if (medium) {
      return FontFamily.medium
    }
    if (regular) {
      return FontFamily.regular
    }
  }, [extraBold, bold, semiBold, medium, regular])

  const rightIndicator = useMemo(() => {
    if (error) {
      return <InvalidIndicator />
    }

    if (isValid) {
      return <ValidIndicator />
    }
  }, [error, isValid])

  const [focused, setFocused] = useState(false)

  const handleFocus = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    setFocused(true)
    onFocus?.(e)
  }

  const handleBlur = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    setFocused(false)
    onBlur?.(e)
  }

  return (
    <>
      <View style={containerStyle}>
        {!!label && (
          <Text testID={labelTestID} color={colors.mono100} marginBottom={8}>
            {label}
          </Text>
        )}
        <Box
          borderRadius={8}
          borderColor={focused ? colors.primary80 : colors.mono10}
          backgroundColor={editable === false ? colors.mono5 : colors.white}
          borderWidth={1}
          flexDirection="row"
          alignItems="center"
          width="100%"
          height={48}
          paddingVertical={INPUT_PADDING_VERTICAL}>
          <RNTextInput
            {...inputProps}
            allowFontScaling={false}
            editable={editable}
            onFocus={handleFocus}
            onBlur={handleBlur}
            placeholderTextColor={
              inputProps.placeholderTextColor || colors.mono40
            }
            style={[
              {
                fontFamily,
                fontSize,
                lineHeight,
                color,
                paddingHorizontal: INPUT_PADDING_HORIZONTAL
              },
              {
                flexGrow: 1,
                flexShrink: 1,
                padding: 0
              },
              inputStyle
            ]}
          />

          {!!rightIndicator && (
            <Box paddingRight={INPUT_PADDING_HORIZONTAL}>{rightIndicator}</Box>
          )}
        </Box>
      </View>

      {!!error && (
        <Text testID={msgTestID} marginTop={12} color={colors.destructive}>
          {error}
        </Text>
      )}
    </>
  )
}
