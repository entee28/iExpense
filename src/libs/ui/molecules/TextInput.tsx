import React, { useMemo, useState } from 'react'
import {
  StyleProp,
  TextInputProps,
  TextStyle,
  View,
  ViewStyle,
  TextInput as RNTextInput,
  NativeSyntheticEvent,
  TextInputFocusEventData
} from 'react-native'
import { Box, Text, FontFamily } from '../atoms'
import { InvalidIndicator, ValidIndicator } from 'assets/icons'
import colors from '../colors'

type Props = {
  left?: JSX.Element
  right?: JSX.Element
  error?: string
  isValid?: boolean
  containerStyle?: ViewStyle
  inputStyle?: StyleProp<TextStyle>
  label?: string | JSX.Element
  labelTestID?: string
  msgTestID?: string
} & FontWeight &
  TextInputProps &
  Pick<TextStyle, 'fontSize' | 'lineHeight' | 'color'>

const INPUT_PADDING_HORIZONTAL = 16
const INPUT_PADDING_VERTICAL = 12

export const TextInput = ({
  left,
  right,
  error,
  isValid,
  containerStyle,
  inputStyle,
  editable,
  label,
  bold,
  semiBold,
  medium,
  extraBold,
  regular,
  fontSize,
  lineHeight,
  color,
  onFocus,
  onBlur,
  msgTestID,
  labelTestID,
  ...inputProps
}: Props) => {
  const fontFamily = useMemo(() => {
    if (extraBold) {
      return FontFamily.extraBold
    }
    if (bold) {
      return FontFamily.bold
    }
    if (semiBold) {
      return FontFamily.semiBold
    }
    if (medium) {
      return FontFamily.medium
    }
    if (regular) {
      return FontFamily.regular
    }
  }, [extraBold, bold, semiBold, medium, regular])

  const rightIndicator = useMemo(() => {
    if (error) {
      return <InvalidIndicator testID="invalid_indicator" />
    }

    if (isValid) {
      return <ValidIndicator testID="valid_indicator" />
    }
  }, [error, isValid])

  const [focused, setFocused] = useState(false)

  const handleFocus = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    setFocused(true)
    onFocus?.(e)
  }

  const handleBlur = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    setFocused(false)
    onBlur?.(e)
  }

  return (
    <>
      <View style={containerStyle}>
        {!!label && (
          <Text testID={labelTestID} color={colors.mono100} marginBottom={8}>
            {label}
          </Text>
        )}
        <Box
          borderRadius={8}
          borderColor={focused ? colors.primary80 : colors.mono10}
          backgroundColor={editable === false ? colors.mono5 : colors.white}
          borderWidth={1}
          flexDirection="row"
          alignItems="center"
          width="100%"
          height={48}
          paddingVertical={INPUT_PADDING_VERTICAL}
          testID="text_input.container">
          {!!left && <Box paddingLeft={INPUT_PADDING_HORIZONTAL}>{left}</Box>}

          <RNTextInput
            {...inputProps}
            allowFontScaling={false}
            editable={editable}
            onFocus={handleFocus}
            onBlur={handleBlur}
            placeholderTextColor={
              inputProps.placeholderTextColor || colors.mono40
            }
            style={[
              {
                fontFamily,
                fontSize,
                lineHeight,
                color,
                paddingHorizontal: INPUT_PADDING_HORIZONTAL
              },
              {
                flexGrow: 1,
                flexShrink: 1,
                padding: 0
              },
              inputStyle
            ]}
          />

          {!!right && (
            <Box paddingRight={INPUT_PADDING_HORIZONTAL}>{right}</Box>
          )}

          {!!rightIndicator && (
            <Box paddingRight={INPUT_PADDING_HORIZONTAL}>{rightIndicator}</Box>
          )}
        </Box>
      </View>

      {!!error && (
        <Text testID={msgTestID} marginTop={12} color={colors.destructive}>
          {error}
        </Text>
      )}
    </>
  )
}
