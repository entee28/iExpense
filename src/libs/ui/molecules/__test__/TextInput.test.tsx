import { render } from '@testing-library/react-native'
import colors from 'libs/ui/colors'
import { FontFamily, Text } from '../../atoms'
import { TextInput } from '../TextInput'

it('correctly renders label', () => {
  const textInput = render(
    <TextInput labelTestID="input_label" label="Hello" />
  )

  expect(textInput.getByTestId('input_label').children[0]).toEqual('Hello')
})

it('correctly handles error message', () => {
  const textInput = render(
    <TextInput msgTestID="error_msg" error="Hello Error" />
  )

  expect(textInput.getByTestId('error_msg').children[0]).toEqual('Hello Error')
  expect(textInput.getByTestId('error_msg')).toHaveStyle({
    color: colors.destructive
  })
})

it('should have correct font', () => {
  const regularTextInput = render(<TextInput testID="regular_input" regular />)
  const boldTextInput = render(<TextInput testID="bold_input" bold />)
  const extraBoldTextInput = render(
    <TextInput testID="extra_bold_input" extraBold />
  )
  const mediumTextInput = render(<TextInput testID="medium_input" medium />)
  const semiBoldTextInput = render(
    <TextInput testID="semi_bold_input" semiBold />
  )

  expect(regularTextInput.getByTestId('regular_input')).toHaveStyle({
    fontFamily: FontFamily.regular
  })
  expect(boldTextInput.getByTestId('bold_input')).toHaveStyle({
    fontFamily: FontFamily.bold
  })
  expect(semiBoldTextInput.getByTestId('semi_bold_input')).toHaveStyle({
    fontFamily: FontFamily.semiBold
  })
  expect(mediumTextInput.getByTestId('medium_input')).toHaveStyle({
    fontFamily: FontFamily.medium
  })
  expect(extraBoldTextInput.getByTestId('extra_bold_input')).toHaveStyle({
    fontFamily: FontFamily.extraBold
  })
})

it('should have correct indicator', () => {
  const validTextInput = render(<TextInput isValid />)
  const invalidTextInput = render(<TextInput error="Error" />)

  expect(validTextInput.getByTestId('valid_indicator')).toBeTruthy()

  expect(invalidTextInput.getByTestId('invalid_indicator')).toBeTruthy()
})

it('should render left component if left prop is defined', () => {
  const textInputWithLeftComponent = render(
    <TextInput left={<Text>Hello</Text>} />
  )

  expect(textInputWithLeftComponent.getByText('Hello')).toBeTruthy()
})

it('should have correct background color based on editable prop', () => {
  const editableTextInput = render(<TextInput testID="editable_input" />)

  const uneditableTextInput = render(
    <TextInput editable={false} testID="uneditable_input" />
  )

  expect(editableTextInput.getByTestId('text_input.container')).toHaveStyle({
    backgroundColor: colors.white
  })
  expect(uneditableTextInput.getByTestId('text_input.container')).toHaveStyle({
    backgroundColor: colors.mono5
  })
})

it('renders correctly', () => {
  const snapshot = render(<TextInput value="test" label="hello" />).toJSON()

  expect(snapshot).toMatchSnapshot()
})
