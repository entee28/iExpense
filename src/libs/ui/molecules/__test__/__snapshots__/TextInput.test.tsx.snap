// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <Text
    style={
      [
        {
          "color": "#1A1A1A",
          "fontFamily": "Nunito-Regular",
          "fontSize": 14,
          "lineHeight": undefined,
          "textAlign": "auto",
        },
        {
          "marginBottom": 8,
        },
      ]
    }
  >
    hello
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#1A1A1A1A",
        "borderRadius": 8,
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 48,
        "paddingVertical": 12,
        "width": "100%",
      }
    }
  >
    <TextInput
      allowFontScaling={false}
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#1A1A1A66"
      style={
        [
          {
            "color": undefined,
            "fontFamily": undefined,
            "fontSize": undefined,
            "lineHeight": undefined,
            "paddingHorizontal": 16,
          },
          {
            "flexGrow": 1,
            "flexShrink": 1,
            "padding": 0,
          },
          undefined,
        ]
      }
      value="test"
    />
  </View>
</View>
`;
